Nodejs/Product/TestAdapterImpl/TestContainerDiscoverer.cs

```diff
@@ -99,37 +99,51 @@ internal bool IsTestFile(string pathToFile)
                     return false;
                 }
             }
-            else if (!StringComparer.OrdinalIgnoreCase.Equals(Path.GetExtension(pathToFile), NodejsConstants.JavaScriptExtension))
+            else if (!TypeScriptHelpers.IsJavaScriptFile(pathToFile))
             {
                 return false;
             }
 
-            if(!(project is IVsBuildPropertyStorage propStore))
+            if (TryGetProjectUnitTestProperties(project, out var testRoot, out _) && !string.IsNullOrEmpty(testRoot))
             {
-                Debug.Fail($"Why is {nameof(project)} not of type {nameof(IVsBuildPropertyStorage)}?");
+                project.TryGetProjectDirectory(out var root);
+                var testRootPath = Path.Combine(root, testRoot);
 
-                return false;
+                return CommonUtils.IsSubpathOf(testRootPath, pathToFile);
             }
 
-            var hr = propStore.GetPropertyValue(NodeProjectProperty.TestRoot,/*configuration*/ "", (uint)_PersistStorageType.PST_PROJECT_FILE, out var testRoot);
+            ErrorHandler.Succeeded(((IVsHierarchy)project).ParseCanonicalName(pathToFile, out var itemId));
+
+            return IsTestFile(itemId, project);
+        }
 
-            // if test root is specified check if the file is contained, otherwise fall back to old logic
-            if (ErrorHandler.Succeeded(hr) && !string.IsNullOrEmpty(testRoot))
+        private static bool TryGetProjectUnitTestProperties(IVsProject project, out string testRoot, out string testFramework)
+        {
+            if (!(project is IVsBuildPropertyStorage propStore))
             {
-                project.TryGetProjectPath(out var root);
-                var testRootPath = Path.Combine(root, testRoot);
+                Debug.Fail($"Why is {nameof(project)} not of type {nameof(IVsBuildPropertyStorage)}?");
+
+                testRoot = null;
+                testFramework = null;
 
-                return CommonUtils.IsSubpathOf(root, pathToFile);
+                return false;
             }
 
-            ErrorHandler.Succeeded(((IVsHierarchy)project).ParseCanonicalName(pathToFile, out var itemId));
+            // .NET Framework returns error, .NET Core returns success with "" as output, NTVS returns success with null as output
+            var hrTestRoot = propStore.GetPropertyValue(NodeProjectProperty.TestRoot, /*configuration*/"", (uint)_PersistStorageType.PST_PROJECT_FILE, out testRoot);
+            var hrTestFramework = propStore.GetPropertyValue(NodeProjectProperty.TestFramework, /*configuration*/"", (uint)_PersistStorageType.PST_PROJECT_FILE, out testFramework);
 
-            return IsTestFile(itemId, project);
+            if (ErrorHandler.Failed(hrTestRoot) && ErrorHandler.Failed(hrTestFramework)
+                || testRoot == string.Empty && testFramework == string.Empty)
```

Comment:
I'm not sure if I understand your question. If any of the properties is found this method will return `true`. Only when everything has either failed or all properties are `string.Empty` will return false. 

Note that properties with `null` is a success scenario.

Hope this makes sense.

Reply:
What happens if `hrTestRoot` indicates failure, `hrTestFramework` indicates success, and `testFramework == string.Empty`?

Reply:
That scenario should only happen for the .NET Framework if the user has configured incorrectly the project file. I'd agree that it should display some kind of error message. Unfortunately since the only easy way to really detect if the setting is setup is by actually setting a value this is a drawback I'm willing to accept. 

In any case, an empty string on any of the properties on a .NET Framework project will be an issue for the user that will not allow the test to be run at all.

Let me know if you have a better approach to this.

Reply:
Seems like you want `(ErrorHandler.Failed(hrTestRoot) || testRoot == string.Empty) && (ErrorHandler.Failed(hrTestFramework) || testFramework == string.Empty)`.

Reply:
Offline, you explained that the meaning of empty string depends on the project system.